## DataSet-Link : http://www2.informatik.uni-freiburg.de/~cziegler/BX/

In [1]:
import numpy as np
import pandas as pd

In [2]:
books = pd.read_csv('BX-Books.csv',sep=";",error_bad_lines=False,encoding='latin-1')

b'Skipping line 6452: expected 8 fields, saw 9\nSkipping line 43667: expected 8 fields, saw 10\nSkipping line 51751: expected 8 fields, saw 9\n'
b'Skipping line 92038: expected 8 fields, saw 9\nSkipping line 104319: expected 8 fields, saw 9\nSkipping line 121768: expected 8 fields, saw 9\n'
b'Skipping line 144058: expected 8 fields, saw 9\nSkipping line 150789: expected 8 fields, saw 9\nSkipping line 157128: expected 8 fields, saw 9\nSkipping line 180189: expected 8 fields, saw 9\nSkipping line 185738: expected 8 fields, saw 9\n'
b'Skipping line 209388: expected 8 fields, saw 9\nSkipping line 220626: expected 8 fields, saw 9\nSkipping line 227933: expected 8 fields, saw 11\nSkipping line 228957: expected 8 fields, saw 10\nSkipping line 245933: expected 8 fields, saw 9\nSkipping line 251296: expected 8 fields, saw 9\nSkipping line 259941: expected 8 fields, saw 9\nSkipping line 261529: expected 8 fields, saw 9\n'
C:\Users\user\anaconda3\lib\site-packages\IPython\core\interactiveshell.py

In [3]:
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [4]:
books.columns

Index(['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher',
       'Image-URL-S', 'Image-URL-M', 'Image-URL-L'],
      dtype='object')

In [5]:
# Selcting only required columns
books = books[['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher']]

In [6]:
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  
0    Mark P. O. Morford                2002     Oxford University Press  
1  Richard Bruce Wright                2001       HarperFlamingo Canada  
2          Carlo D'Este                1991             HarperPerennial  
3      Gina Bari Kolata                1999        Farrar Straus Giroux  
4       E. J. W. Barber                1999  W. W. Norton &amp; Company

In [7]:
# Renaming columns as per our usage
books.rename(columns={'Book-Title':'title','Book-Author':'author','Year-Of-Publication':'year','Publisher':'publisher'},inplace=True)

In [8]:
books.head()

ISBN                                              title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

                 author  year                   publisher  
0    Mark P. O. Morford  2002     Oxford University Press  
1  Richard Bruce Wright  2001       HarperFlamingo Canada  
2          Carlo D'Este  1991             HarperPerennial  
3      Gina Bari Kolata  1999        Farrar Straus Giroux  
4       E. J. W. Barber  1999  W. W. Norton &amp; Company

## Selcting Second DataFrame-Users

In [9]:
users = pd.read_csv('BX-Users.csv',sep=";",error_bad_lines=False,encoding='latin-1')
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [10]:
users.rename(columns={'User-ID':'user_id','Location':'location','Age':'age'})

user_id                            location   age
0             1                  nyc, new york, usa   NaN
1             2           stockton, california, usa  18.0
2             3     moscow, yukon territory, russia   NaN
3             4           porto, v.n.gaia, portugal  17.0
4             5  farnborough, hants, united kingdom   NaN
...         ...                                 ...   ...
278853   278854               portland, oregon, usa   NaN
278854   278855  tacoma, washington, united kingdom  50.0
278855   278856           brampton, ontario, canada   NaN
278856   278857           knoxville, tennessee, usa   NaN
278857   278858                dublin, n/a, ireland   NaN

[278858 rows x 3 columns]

## Adding Rating data Set

In [11]:
ratings = pd.read_csv('BX-Book-Ratings.csv',sep=';',error_bad_lines=False,encoding='latin-1')
ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [12]:
ratings.rename(columns={'User-ID':'user_id','Book-Rating':'rating'},inplace=True)
ratings.head()

user_id        ISBN  rating
0   276725  034545104X       0
1   276726  0155061224       5
2   276727  0446520802       0
3   276729  052165615X       3
4   276729  0521795028       6

In [13]:
books.shape

(271360, 5)

In [14]:
users.shape

(278858, 3)

In [15]:
ratings.shape

(1149780, 3)

## Here we were going to use the person  who had read more than 200 books
## And we were taking the person who had given more 50 ratings 

* Rating columns DataSet

In [16]:
# Caliculating each person revied the books
ratings['user_id'].value_counts()

11676     13602
198711     7550
153662     6109
98391      5891
35859      5850
          ...  
271728        1
245123        1
234886        1
259466        1
187812        1
Name: user_id, Length: 105283, dtype: int64

In [17]:
# Number of revied persons
ratings['user_id'].value_counts().shape

(105283,)

In [18]:
# Checking the person who revied more than 200 books
x = ratings['user_id'].value_counts()>200

In [19]:
x.shape

(105283,)

In [20]:
# No of persons readed more than 200 reviewd
y = x[x].index

In [21]:
y.shape

(899,)

In [22]:
y

Int64Index([ 11676, 198711, 153662,  98391,  35859, 212898, 278418,  76352,
            110973, 235105,
            ...
            116122,  28634, 188951,  59727, 155916, 274808,  73681,   9856,
            268622,  44296],
           dtype='int64', length=899)

In [23]:
# Creating a data frame those readed more than 200 books
ratings = ratings[ratings['user_id'].isin(y)]
ratings.head()

user_id        ISBN  rating
1456   277427  002542730X      10
1457   277427  0026217457       0
1458   277427  003008685X       8
1459   277427  0030615321       0
1460   277427  0060002050       0

In [24]:
# Total no of Ratings done by this people
ratings.shape

(526356, 3)

* Out of 105283 ratings 526356 ratings are done by this people reamainin done by others

In [25]:
books.head(2)

ISBN                title                author  year  \
0  0195153448  Classical Mythology    Mark P. O. Morford  2002   
1  0002005018         Clara Callan  Richard Bruce Wright  2001   

                 publisher  
0  Oxford University Press  
1    HarperFlamingo Canada

In [26]:
ratings.head(2)

user_id        ISBN  rating
1456   277427  002542730X      10
1457   277427  0026217457       0

In [27]:
books.shape

(271360, 5)

In [28]:
ratings.shape

(526356, 3)

In [29]:
# Merge this books data set to the ratings columns
ratings_with_books = ratings.merge(books,on='ISBN')

In [30]:
ratings_with_books

user_id        ISBN  rating  \
0        277427  002542730X      10   
1          3363  002542730X       0   
2         11676  002542730X       6   
3         12538  002542730X      10   
4         13552  002542730X       0   
...         ...         ...     ...   
487666   275970  1892145022       0   
487667   275970  1931868123       0   
487668   275970  3411086211      10   
487669   275970  3829021860       0   
487670   275970  4770019572       0   

                                                    title  \
0       Politically Correct Bedtime Stories: Modern Ta...   
1       Politically Correct Bedtime Stories: Modern Ta...   
2       Politically Correct Bedtime Stories: Modern Ta...   
3       Politically Correct Bedtime Stories: Modern Ta...   
4       Politically Correct Bedtime Stories: Modern Ta...   
...                                                   ...   
487666                                   Here Is New York   
487667  There's a Porcupine in My Outhouse: Misadventu...   
487668                                         Die Biene.   
487669                                     The Penis Book   
487670                                            Musashi   

                               author  year  \
0                   James Finn Garner  1994   
1                   James Finn Garner  1994   
2                   James Finn Garner  1994   
3                   James Finn Garner  1994   
4                   James Finn Garner  1994   
...                               ...   ...   
487666                    E. B. White  1999   
487667                   Mike Tougias  2002   
487668  Sybil Gr�¤fin Sch�¶nfeldt  1993   
487669                   Joseph Cohen  1999   
487670                 Eiji Yoshikawa  1995   

                                   publisher  
0                  John Wiley &amp; Sons Inc  
1                  John Wiley &amp; Sons Inc  
2                  John Wiley &amp; Sons Inc  
3                  John Wiley &amp; Sons Inc  
4                  John Wiley &amp; Sons Inc  
...                                      ...  
487666                       Little Bookroom  
487667                    Capital Books (VA)  
487668  Bibliographisches Institut, Mannheim  
487669                              Konemann  
487670          Kodansha International (JPN)  

[487671 rows x 7 columns]

In [31]:
# Checking the reviews of the each book
number_rating = ratings_with_books.groupby('title')['rating'].count().reset_index()

In [32]:
number_rating.rename(columns={'rating':'number of ratings'},inplace=True)

In [33]:
number_rating.head()

title  number of ratings
0   A Light in the Storm: The Civil War Diary of ...                  2
1                              Always Have Popsicles                  1
2               Apple Magic (The Collector's series)                  1
3   Beyond IBM: Leadership Marketing and Finance ...                  1
4   Clifford Visita El Hospital (Clifford El Gran...                  1

In [34]:
# Mearging this dataset on the ratings with books data set
final_rating = ratings_with_books.merge(number_rating, on='title')
final_rating

user_id        ISBN  rating  \
0        277427  002542730X      10   
1          3363  002542730X       0   
2         11676  002542730X       6   
3         12538  002542730X      10   
4         13552  002542730X       0   
...         ...         ...     ...   
487666   275970  1892145022       0   
487667   275970  1931868123       0   
487668   275970  3411086211      10   
487669   275970  3829021860       0   
487670   275970  4770019572       0   

                                                    title  \
0       Politically Correct Bedtime Stories: Modern Ta...   
1       Politically Correct Bedtime Stories: Modern Ta...   
2       Politically Correct Bedtime Stories: Modern Ta...   
3       Politically Correct Bedtime Stories: Modern Ta...   
4       Politically Correct Bedtime Stories: Modern Ta...   
...                                                   ...   
487666                                   Here Is New York   
487667  There's a Porcupine in My Outhouse: Misadventu...   
487668                                         Die Biene.   
487669                                     The Penis Book   
487670                                            Musashi   

                               author  year  \
0                   James Finn Garner  1994   
1                   James Finn Garner  1994   
2                   James Finn Garner  1994   
3                   James Finn Garner  1994   
4                   James Finn Garner  1994   
...                               ...   ...   
487666                    E. B. White  1999   
487667                   Mike Tougias  2002   
487668  Sybil Gr�¤fin Sch�¶nfeldt  1993   
487669                   Joseph Cohen  1999   
487670                 Eiji Yoshikawa  1995   

                                   publisher  number of ratings  
0                  John Wiley &amp; Sons Inc                 82  
1                  John Wiley &amp; Sons Inc                 82  
2                  John Wiley &amp; Sons Inc                 82  
3                  John Wiley &amp; Sons Inc                 82  
4                  John Wiley &amp; Sons Inc                 82  
...                                      ...                ...  
487666                       Little Bookroom                  1  
487667                    Capital Books (VA)                  1  
487668  Bibliographisches Institut, Mannheim                  1  
487669                              Konemann                  1  
487670          Kodansha International (JPN)                  1  

[487671 rows x 8 columns]

### Accepting more than 50 datasets we selected

In [35]:
final_rating = final_rating[final_rating['number of ratings']>=50]
final_rating

user_id        ISBN  rating  \
0        277427  002542730X      10   
1          3363  002542730X       0   
2         11676  002542730X       6   
3         12538  002542730X      10   
4         13552  002542730X       0   
...         ...         ...     ...   
236701   255489  0553579983       7   
236702   256407  0553579983       0   
236703   257204  0553579983       0   
236704   261829  0553579983       0   
236705   273979  0553579983       0   

                                                    title             author  \
0       Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
1       Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
2       Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
3       Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
4       Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
...                                                   ...                ...   
236701                                   And Then You Die      Iris Johansen   
236702                                   And Then You Die      Iris Johansen   
236703                                   And Then You Die      Iris Johansen   
236704                                   And Then You Die      Iris Johansen   
236705                                   And Then You Die      Iris Johansen   

        year                  publisher  number of ratings  
0       1994  John Wiley &amp; Sons Inc                 82  
1       1994  John Wiley &amp; Sons Inc                 82  
2       1994  John Wiley &amp; Sons Inc                 82  
3       1994  John Wiley &amp; Sons Inc                 82  
4       1994  John Wiley &amp; Sons Inc                 82  
...      ...                        ...                ...  
236701  1998                     Bantam                 50  
236702  1998                     Bantam                 50  
236703  1998                     Bantam                 50  
236704  1998                     Bantam                 50  
236705  1998                     Bantam                 50  

[61853 rows x 8 columns]

* We need to check the same person on a single book with multiple time need to remove

In [36]:
final_rating

user_id        ISBN  rating  \
0        277427  002542730X      10   
1          3363  002542730X       0   
2         11676  002542730X       6   
3         12538  002542730X      10   
4         13552  002542730X       0   
...         ...         ...     ...   
236701   255489  0553579983       7   
236702   256407  0553579983       0   
236703   257204  0553579983       0   
236704   261829  0553579983       0   
236705   273979  0553579983       0   

                                                    title             author  \
0       Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
1       Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
2       Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
3       Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
4       Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
...                                                   ...                ...   
236701                                   And Then You Die      Iris Johansen   
236702                                   And Then You Die      Iris Johansen   
236703                                   And Then You Die      Iris Johansen   
236704                                   And Then You Die      Iris Johansen   
236705                                   And Then You Die      Iris Johansen   

        year                  publisher  number of ratings  
0       1994  John Wiley &amp; Sons Inc                 82  
1       1994  John Wiley &amp; Sons Inc                 82  
2       1994  John Wiley &amp; Sons Inc                 82  
3       1994  John Wiley &amp; Sons Inc                 82  
4       1994  John Wiley &amp; Sons Inc                 82  
...      ...                        ...                ...  
236701  1998                     Bantam                 50  
236702  1998                     Bantam                 50  
236703  1998                     Bantam                 50  
236704  1998                     Bantam                 50  
236705  1998                     Bantam                 50  

[61853 rows x 8 columns]

In [37]:
final_rating.shape

(61853, 8)

In [38]:
final_rating.drop_duplicates(['user_id','title'],inplace=True)

<ipython-input-38-88ef78dad5f7>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_rating.drop_duplicates(['user_id','title'],inplace=True)


In [39]:
final_rating.shape

(59850, 8)

# Creating a Pivot Table

In [40]:
book_pivot = final_rating.pivot_table(columns='user_id',index='title',values='rating')
book_pivot

user_id                                             254     2276    2766    \
title                                                                        
1984                                                   9.0     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN    10.0     NaN   
4 Blondes                                              NaN     NaN     NaN   
84 Charing Cross Road                                  NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

user_id                                             2977    3363    3757    \
title                                                                        
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
84 Charing Cross Road                                  NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

user_id                                             4017    4385    6242    \
title                                                                        
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
84 Charing Cross Road                                  NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     7.0   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

user_id                                             6251    ...  274004  \
title                                                       ...           
1984                                                   NaN  ...     NaN   
1st to Die: A Novel                                    NaN  ...     NaN   
2nd Chance                                             NaN  ...     NaN   
4 Blondes                                              0.0  ...     NaN   
84 Charing Cross Road                                  NaN  ...     NaN   
...                                                    ...  ...     ...   
Year of Wonders                                        NaN  ...     NaN   
You Belong To Me                                       NaN  ...     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     NaN   
Zoya                                                   NaN  ...     NaN   
\O\" Is for Outlaw"                                    

In [41]:
# Filling missing values 
book_pivot.fillna(0,inplace=True)
book_pivot

user_id                                             254     2276    2766    \
title                                                                        
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             2977    3363    3757    \
title                                                                        
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             4017    4385    6242    \
title                                                                        
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     7.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             6251    ...  274004  \
title                                                       ...           
1984                                                   0.0  ...     0.0   
1st to Die: A Novel                                    0.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
84 Charing Cross Road                                  0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

# Converting into Sparesmmatrix
* This means it wont considerd zero values

In [42]:
from scipy.sparse import csr_matrix
book_sparse = csr_matrix(book_pivot)

In [43]:
type(book_sparse)

scipy.sparse.csr.csr_matrix

## Model deployment
* Clustring Model

In [44]:
from sklearn.neighbors import NearestNeighbors
model = NearestNeighbors(algorithm='brute')

In [45]:
model.fit(book_sparse)

NearestNeighbors(algorithm='brute')

* Slecting single book

In [46]:
# Example of selecting a book number
book_pivot.index[237]

'Harry Potter and the Chamber of Secrets (Book 2)'

In [47]:
distances, suggestions = model.kneighbors(book_pivot.iloc[237,:].values.reshape(1,-1), n_neighbors=6)

In [48]:
distances

array([[ 0.        , 68.78953409, 69.5413546 , 72.64296249, 76.83098333,
        77.28518616]])

* 0. means independent of that book and remaining are the id

In [49]:
suggestions

array([[237, 240, 238, 241, 184, 536]], dtype=int64)

* These are the books that actuaaly recomending

In [50]:
# To display their names
for i in range(len(suggestions)):
    print(book_pivot.index[suggestions[i]])

Index(['Harry Potter and the Chamber of Secrets (Book 2)',
       'Harry Potter and the Prisoner of Azkaban (Book 3)',
       'Harry Potter and the Goblet of Fire (Book 4)',
       'Harry Potter and the Sorcerer's Stone (Book 1)', 'Exclusive',
       'The Cradle Will Fall'],
      dtype='object', name='title')


## By displaying with name we need to crate a helper function

In [51]:
np.where(book_pivot.index == 'Animal Farm')

(array([54], dtype=int64),)

* It displays its id number 

In [52]:
# To convert into frame
np.where(book_pivot.index == 'Animal Farm')[0][0]

54

In [53]:
# To Create Recomend Function
def recommend_book(book_name):
    book_id = np.where(book_pivot.index == book_name)[0][0]
    distances, suggestions = model.kneighbors(book_pivot.iloc[book_id,:].values.reshape(1,-1), n_neighbors=6)
    
    for i in range(len(suggestions)):
        if i==0:
            print("The suggestion for",book_name,'are :')
        if not i:
            print(book_pivot.index[suggestions[i]])

In [54]:
recommend_book('Harry Potter and the Chamber of Secrets (Book 2)')

The suggestion for Harry Potter and the Chamber of Secrets (Book 2) are :
Index(['Harry Potter and the Chamber of Secrets (Book 2)',
       'Harry Potter and the Prisoner of Azkaban (Book 3)',
       'Harry Potter and the Goblet of Fire (Book 4)',
       'Harry Potter and the Sorcerer's Stone (Book 1)', 'Exclusive',
       'The Cradle Will Fall'],
      dtype='object', name='title')
